### PR TITLE
[Bugfix][MiniCPMO45] Return bare tensor in Thinker forward to fix repetition loop (#7497)

### LINKED ISSUE #7497: [Bug]: MiniCPM-o 4.5 stage-0 LLM class returns (input_embeddings, hidden_states) — the AR runner applies the LM head to raw input embeddings, silently degrading every rollout into self-repetition
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Your output of `python collect_env.py` here
```

</details>

### Your code version

<details>
<summary>The commit id or version of vllm</summary>

```text

```
</details>
<details>
<summary>The commit id or version of vllm-omni</summary>

```text

```
</details>

### 🐛 Describe the bug

**Summary.** `MiniCPMO45OmniLLMForConditionalGeneration.forward` returns the raw
tuple `(text_inputs_embeds, hidden_states)` — **input embeddings first**. The AR
runner's `extract_multimodal_outputs` treats any plain tuple return as
`(text_hidden_states, *rest)` and takes **element `[0]`** as the hidden states.
Logits are therefore computed by applying the LM head to the **raw input
embeddings** instead of the transformer output. The embeddings tensor is
exactly the `[num_tokens, hidden]` layout the runner expects, so nothing
crashes or warns — every sampled token
is just a near-deterministic function of the current token's embedding, and
generation degenerates into confident self-repetition loops until the response
cap. The model's real hidden states (element `[1]`) are computed and silently
discarded.

This is the standalone-LLM-class sibling of #7360 (closed by #7384): #7384 fixed
the *input* half (`embed_multimodal`); this is the *output* half, which is still
unfixed on main.

**Where it happens (main @ `58adeec05`).**

Model side — `vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py:4639`:

```python
hidden_states = self.llm.model(input_ids, positions, intermediate_tensors, inputs_embeds=inputs_embeds)
return text_inputs_embeds, hidden_states.unsqueeze(0) if hidden_states.ndim == 2 else hidden_states
```

`text_inputs_embeds` is the raw embedding-layer output — with multimodal spans
zeroed (`torch.zeros_like`) when the class computes embeddings itself
(:4624-4632), or the runner-supplied `inputs_embeds` as-is in the V1
precomputed-embeds branch (:4634). Nothing consumes it by design: its only
caller outside the runner, the 3-stage wrapper, binds it at
`minicpmo_4_5_omni.py:550` and never uses it — a dead payload the AR runner
mis-takes as the hidden states.

Runner side — `vllm_omni/worker/gpu_model_runner.py`, `extract_multimodal_outputs`:

```python
elif isinstance(hidden_states, list) or isinstance(hidden_states, tuple):
    text_hidden_states = hidden_states[0]      # ← picks text_inputs_embeds
    multimodal_outputs = {}
```

called from `vllm_omni/worker/gpu_ar_model_runner.py:1295` in `execute_model`,
after which `:1339-1346` does `compute_logits(hidden_states[logits_indices])`.

The runner's convention is sound and matches stock vLLM (aux-output tuples are
unpacked `hidden_states, aux_hidden_states = model_output`, e.g. the EAGLE-3
path in vLLM's `GPUModelRunner`), and vllm-omni's own `OmniOutput` is unpacked
by field name, which is immune. `MiniCPMO45OmniLLM` is the one registered AR
model that returns a *payload-first* raw tuple on the default path.

**Reproduce.** Stand up any engine whose stage-0 model resolves to
`MiniCPMO45OmniLLMForConditionalGeneration` (single-stage/thinker-only deploy —
e.g. a stage with `model_arch: MiniCPMO45OmniLLMForConditionalGeneration` in the
deploy config / `OmniEngineArgs`, which injects it into
`hf_overrides.architectures`, `vllm_omni/engine/arg_utils.py:317-321`). Then any
generation request on an image+audio prompt returns the degenerate output below.
The bug sits below the request API, so every entrypoint converges on it —
`AsyncOmni.generate`, the sync `Omni` class, and `vllm serve`'s OpenAI
endpoints all reach the same path:

```text
AsyncOmni.generate / Omni.generate / vllm serve
  → EngineCore → stage-0 worker (GPUARWorker, gpu_ar_worker.py:31)
    → GPUARModelRunner.execute_model            (gpu_ar_model_runner.py)
        → model.forward  →  (text_inputs_embeds, hidden_states)   ← :4639
        → extract_multimodal_outputs(model_output) → takes [0]    ← :1295
        → compute_logits(input_embeddings)                         ← :1342
        → sampler → self-repetition
```

All our evidence was collected through `AsyncOmni.generate`
(`vllm_omni/entrypoints/async_omni.py:464`), the entrypoint an RL rollout loop
drives (verl-omni's `ARStrategy.run_generation` calls
`self.server.engine.generate(...)` with `prompt` / `sampling_params_list` /
`request_id` and consumes the async generator). Minimal shape:

```python
from vllm_omni.entrypoints import AsyncOmni

engine = AsyncOmni(model="/path/to/MiniCPM-o-4_5",
                   deploy_config="<single-stage AR deploy config you define — the stock "
                                  "minicpmo_4_5 yamls are all 3-stage; set the stage's "
                                  "model_arch to MiniCPMO45OmniLLMForConditionalGeneration>")
async for output in engine.generate(prompt={"prompt": "...", "multi_modal_data": {...}},
                                    sampling_params_list=[...], request_id="r0"):
    ...
```

Observed (before fix) vs expected (after normalizing the forward return to a
bare tensor):

- ~a few coherent tokens, then a high-confidence repetition loop
  (`"The video, I, I, I, …"`, `"and and and"`) with sampled-token logprobs ≈ 0
  (p≈1) inside the loop, truncated at `max_tokens`;
- expected: a short, media-grounded answer with `finish_reason: stop`
  (41-52 tokens on the same prompts once the return value is fixed).

Observed during an RL bring-up against this class: 98-100% of rollouts hit the
response cap; after normalizing the forward return, the same prompts, weights,
and sampling produced correct `<answer>…</answer>` outputs matching ground truth.
The same checkpoint/prompt through the transformers remote code generates a
correct answer (89 tokens, clean stop), so this is a vllm-omni-only defect, not
a transformers/model issue.

The mechanism is verifiable statically — no GPU needed to see the mismatch:

```python
# vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py:4639
return text_inputs_embeds, hidden_states...        # embeddings at [0]
# vllm_omni/worker/gpu_model_runner.py (extract_multimodal_outputs)
text_hidden_states = hidden_states[0]              # runner takes [0]
```

**Scope — which models are affected** (survey of every runner-facing `forward`
on main @ `58adeec05`):

| Model | Forward return | Affected? |
|---|---|---|
| MiniCPM-o 4.5 LLM (`MiniCPMO45OmniLLMForConditionalGeneration`) | raw tuple, **embeddings first**, all paths | **yes — active** |
| MiniCPM-o 4.5 full 3-stage wrapper | unpacks thinker tuple positionally, returns `OmniOutput` (`minicpmo_4_5_omni.py:549-550,610`) | no |
| Qwen3-Omni MoE thinker | bare tensor by default; `(hidden_states, captured)` only on staged capture — hidden at `[0]`, with an explicit comment about exactly this trap (`qwen3_omni_moe_thinker.py:1393-1400`) | no |
| Qwen2.5-Omni thinker, GLM-image (inner), hunyuan vendored, other standard LLM/thinker classes | bare tensor, `OmniOutput`, or hidden-at-`[0]` tuples | no |
| MiMo-audio LLM (`MiMoAudioLLMForConditionalGeneration`) | raw tuple **payload-first** `(batch_next_speech_tokens, hidden_states)` (`mimo_audio_llm.py:1192`), and `[0]` is `None` when no speech tokens are generated | **latent** — registered standalone (`MiMoAudioLLMModel`) but today only consumed via the wrapper, which unpacks positionally (`mimo_audio.py:791`). A standalone user would hit the same mismatch (loudly — `None` crashes rather than corrupts). |

**Suggested fix.** Make `MiniCPMO45OmniLLMForConditionalGeneration.forward`
return a bare `[num_tokens, hidden]` tensor — the same pattern as the Qwen3-Omni
thinker. The embeds element is dead payload even for the 3-stage wrapper (bound
at `minicpmo_4_5_omni.py:550`, never read), so it can simply be dropped; if a
future stage needs it, that is what `OmniOutput.multimodal_outputs` is for.
Consider also
documenting or making loud the raw-tuple branch of
`extract_multimodal_outputs` (e.g. log once when a non-`OmniOutput` tuple is
consumed), since the ordering contract is currently implicit; that would also
catch the MiMo latent case if it is ever run standalone.

**Workaround we ship meanwhile** (verl-omni side, happy to retire it once this
lands): a `vllm.general_plugins` plugin wraps the class's forward and returns
element `[1]` squeezed to `[num_tokens, hidden]`; plain-tensor returns (what a
fixed upstream produces) pass through untouched
(`verl_omni/vllm_plugins.py::_normalize_forward_return` in
verl-project/verl-omni#572).

**Related.**

- #7360 / #7384 — input half (`embed_multimodal` alias) on the same class; fixed.
- verl-project/verl-omni#572 — MiniCPM-o 4.5 thinker-only GSPO recipe; its
  entire zero-reward bring-up traced to this bug (root-cause record in the PR
  description).

Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by BeatSeat**
Fixes #7497.

## Purpose
In `MiniCPMO45OmniLLMForConditionalGeneration.forward()`, the method was returning a tuple `(text_inputs_embeds, hidden_states...)` and computing `text_inputs_embeds` which was dead code. 

- In `vllm-omni`'s AR model runner (`GPUARModelRunner`), `extract_multimodal_outputs()` unpacked `hidden_states[0]`, mistakenly selecting raw `text_inputs_embeds` instead of final hidden states and computing logits on raw embeddings, degrading autoregressive rollouts into infinite self-repetition loops (#7497).
- In stock vLLM `GPUModelRunner.execute_model()`, model forward output is expected to be a bare `torch.Tensor` (`assert isinstance(model_output, torch.Tensor)`). Returning a tuple caused standalone generation to crash with an `AssertionError`.

This PR:
1. Normalizes `MiniCPMO45OmniLLMForConditionalGeneration.forward()` to return bare `hidden_states` `torch.Tensor`, removing the unused `text_inputs_embeds` computation.
2. Updates `minicpmo_4_5_omni.py` to only squeeze dimension 0 when `ndim == 3 and shape[0] == 1`, ensuring consistent 2D `(seq_len, hidden_dim)` hidden states and preserving 2D `(1, H)` single-token decode states.
3. Adds unit tests covering:
   - Bare hidden states return type and values;
   - Single-pass embedding construction when `inputs_embeds=None`;
   - Wrapper shape normalization across `(1, H)`, `(N, H)`, `(1, 1, H)`, and `(1, N, H)`.

## Test Plan

**vLLM Version:** 0.29.0

**vLLM-Omni Commit:** 2cbfd3e93

## Test Result

A/B standalone inference comparison on `MiniCPM-o-4_5` (NVIDIA A800 80GB GPU):

### Case A (Baseline / Unpatched Bug Reproduction)
When logits are computed on raw input embeddings (simulating `extract_multimodal_outputs()` selecting element `[0]`):
```text
================ RESULTS ================
Prompt 1: 'The video, I'
Generated: '. I. I. I. I. I. I. I. I. I. I. I. I. I. I. I'
Tokens: [13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358, 13, 358]
Finish reason: length
------------------------------------------------------------
Prompt 2: 'Describe the scene and'
Generated: '.  .  .  .  .  .  .  .  .  .  '
Tokens: [13, 220, 220, 13, 220, 220, 13, 220, 220, 13, 220, 220, 13, 220, 220, 13, 220, 220, 13, 220, 220, 13, 220, 220, 13, 220, 220, 13, 220, 220]
Finish reason: length
```
*(Direct raw tuple output also failed in stock vLLM with `assert isinstance(model_output, torch.Tensor) AssertionError`)*.

### Case B (Patched / With This Fix)
Inference succeeds, outputs clean completions, and stops normally:
```text
================ RESULTS ================
Prompt 1: 'Question: What is the capital of France? Answer:'
Generated: ' Paris. Category: World Capitals.'
Tokens: [12095, 13, 10054, 25, 4337, 94284, 13, 151645]
Finish reason: stop
------------------------------------------------------------
Prompt 2: 'Question: What is 2 + 2? Answer:'
Generated: ' 4'
Tokens: [220, 19, 151645]
Finish reason: stop
------------------------------------------------------------
```

### Unit Tests
`tests/model_executor/models/minicpmo_4_5/test_minicpmo_4_5_omni_llm_forward.py`:
- `test_thinker_forward_returns_bare_hidden_states`: PASSED
- `test_thinker_forward_builds_embeddings_once_when_inputs_embeds_none`: PASSED
- `test_wrapper_shape_normalization` across `(1, H)`, `(N, H)`, `(1, 1, H)`, and `(1, N, H)`: PASSED

### CODE REVIEW

**Comment by amy-why-3459** [comment]
@zhtmike PTAL

File: vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni.py [hunk 1]
```diff
@@ -547,13 +547,13 @@ def forward(
...
-                embeds, text_hidden_states = thinker_output
+                _, text_hidden_states = thinker_output
             else:
                 text_hidden_states = thinker_output
 
             # Prepare hidden states for downstream stages
-            # Ensure correct shape: (batch_size, seq_len, hidden_dim)
-            if added_batch_dim:
+            # Ensure correct shape: (seq_len, hidden_dim)
+            if text_hidden_states.ndim == 3 and text_hidden_states.shape[0] == 1:
```
Comment (amy-why-3459) [inline vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni.py:556]: Non-blocking: please add a wrapper-level regression test for this shape normalization.

The new test calls only the inner Thinker, so it does not exercise this changed squeeze condition. Parameterize Thinker outputs with shapes `(1, H)`, `(N, H)`, `(1, 1, H)`, and `(1, N, H)`, and verify both `OmniOutput.text_hidden_states` and `multimodal_outputs["latent"]`.

The single-token `(1, H)` case is particularly important: it must remain two-dimensional rather than being squeezed to `(H,)`. I checked these cases against the current wrapper and they pass; committing the coverage would protect this fix.

File: tests/model_executor/models/minicpmo_4_5/test_minicpmo_4_5_omni_llm_forward.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    model = MiniCPMO45OmniLLMForConditionalGeneration.__new__(MiniCPMO45OmniLLMForConditionalGeneration)
+    torch.nn.Module.__init__(model)
+    model.llm = MagicMock()
+    mock_hidden_states = torch.randn(4, 4096)
+    model.llm.model = MagicMock(return_value=mock_hidden_states)
+
+    out = model.forward(
+        input_ids=torch.tensor([1, 2, 3, 4]),
+        positions=torch.tensor([0, 1, 2, 3]),
+        inputs_embeds=torch.randn(4, 4096),
```
Comment (amy-why-3459) [inline tests/model_executor/models/minicpmo_4_5/test_minicpmo_4_5_omni_llm_forward.py:33]: Non-blocking: could we also cover `inputs_embeds=None`?

This test supplies embeddings explicitly, bypassing the branch where the duplicate embedding computation was removed. An additional test can mock `get_multimodal_embeddings()` and `get_input_embeddings()`, assert that input embeddings are built exactly once, and verify that they are forwarded to `self.llm.model` with `input_ids=None`. That would cover the changed construction path as well as the return-value contract.

**Review by amy-why-3459 — APPROVED** [review]
The change is small and consistent with the forward-output contract: return the underlying hidden states directly, remove the unused embedding computation, and squeeze only an actual singleton batch dimension in the wrapper. I found no blocking issues. Two non-blocking test suggestions are inline.

Validation at 63609964374fd2d425155070c7bf1c6e72e3afec: the new unit test passed. I also exercised the real outer wrapper with mocked Thinker outputs shaped (1, H), (N, H), (1, 1, H), and (1, N, H); both text_hidden_states and the latent payload retained the expected two-dimensional shape and values. I did not rerun GPU inference or independently verify the full repetition-loop claim.

**Review by amy-why-3459 — COMMENTED** [review]
Could you please fix the DCO check before merging? Commit `6360996` is missing a `Signed-off-by` trailer, and its author/committer are currently shown as `root`.

Please verify that your Git author identity and sign-off use your correct contributor name and email. If this is your commit and rewriting the PR branch is appropriate, you can amend it on the PR branch and push the update:

```bash
git commit --amend --no-edit --signoff
git push --force-with-lease origin fix/minicpmo45-llm-forward-return-tensor
```

Note that `--signoff` adds the trailer using your configured identity; it does not correct the existing author identity by itself. Please correct the author as well if `root` was accidental, then confirm that DCO passes on the updated commit.

My code approval remains unchanged; this comment is to resolve the outstanding DCO check.

File: vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni.py (hunk 1 above)
Comment (BeatSeat) [reply vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni.py:556]: Thanks for the great suggestion! Added `test_wrapper_shape_normalization` parameterized across `(1, H)`, `(N, H)`, `(1, 1, H)`, and `(1, N, H)` verifying both `out.text_hidden_states` and `out.multimodal_outputs['latent']`, ensuring 2D shapes like `(1, H)` are preserved.

File: tests/model_executor/models/minicpmo_4_5/test_minicpmo_4_5_omni_llm_forward.py (hunk 2 above)
Comment (BeatSeat) [reply tests/model_executor/models/minicpmo_4_5/test_minicpmo_4_5_omni_llm_forward.py:33]: Good point! Added `test_thinker_forward_builds_embeddings_once_when_inputs_embeds_none` which covers the `inputs_embeds=None` path, asserting that `get_input_embeddings` is called exactly once with multimodal embeddings and forwarded to `self.llm.model` with `input_ids=None`.

**Comment by zhtmike** [comment]
> @zhtmike PTAL

LGTM
